src/main/java/com/englishtown/bitbucket/hook/MirrorBucketProcessor.java

```diff
@@ -0,0 +1,147 @@
+
+import com.atlassian.bitbucket.concurrent.BucketProcessor;
+import com.atlassian.bitbucket.i18n.I18nService;
+import com.atlassian.bitbucket.permission.Permission;
+import com.atlassian.bitbucket.repository.Repository;
+import com.atlassian.bitbucket.repository.RepositoryService;
+import com.atlassian.bitbucket.scm.Command;
+import com.atlassian.bitbucket.scm.ScmCommandBuilder;
+import com.atlassian.bitbucket.scm.ScmService;
+import com.atlassian.bitbucket.scm.git.command.GitCommandExitHandler;
+import com.atlassian.bitbucket.server.ApplicationPropertiesService;
+import com.atlassian.bitbucket.user.SecurityService;
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Locale;
+
+import static com.englishtown.bitbucket.hook.MirrorRepositoryHook.PROP_PREFIX;
+
+public class MirrorBucketProcessor implements BucketProcessor<MirrorRequest> {
+
+    static final String PROP_TIMEOUT = PROP_PREFIX + "timeout";
+
+    private static final String DEFAULT_REFSPEC = "+refs/heads/*:refs/heads/*";
+
+    private static final Logger log = LoggerFactory.getLogger(MirrorBucketProcessor.class);
+
+    private final I18nService i18nService;
+    private final PasswordEncryptor passwordEncryptor;
+    private final RepositoryService repositoryService;
+    private final ScmService scmService;
+    private final SecurityService securityService;
+    private final Duration timeout;
+
+    public MirrorBucketProcessor(I18nService i18nService, PasswordEncryptor passwordEncryptor,
+                                 ApplicationPropertiesService propertiesService, RepositoryService repositoryService,
+                                 ScmService scmService, SecurityService securityService) {
+        this.i18nService = i18nService;
+        this.passwordEncryptor = passwordEncryptor;
+        this.repositoryService = repositoryService;
+        this.scmService = scmService;
+        this.securityService = securityService;
+
+        timeout = Duration.ofSeconds(propertiesService.getPluginProperty(PROP_TIMEOUT, 120L));
+    }
+
+    @Override
+    public void process(@Nonnull String key, @Nonnull List<MirrorRequest> requests) {
+        if (requests.isEmpty()) {
+            return;
+        }
+        // Every request is for the same mirror URL, and the same repository ID. In case the
+        // settings (e.g. username/password) have been changed since the first request was
+        // queued, we process the _last_ request in the list. Since mirroring pushes all of
+        // the configured refspecs, any single request should roll up changes from any number
+        // of requests
+        MirrorRequest request = requests.get(requests.size() - 1);
+
+        securityService.withPermission(Permission.REPO_READ, "Mirror changes")
```

Comment:
The `BucketedExecutor` approach needs to use the `SecurityService` to escalate permissions because its executor threads do not preserve the context user--so without escalation they can't access the repository.